### ISSUE TITLE
MapView addressForCoordinate method is throwing error as "Cannot read properties of undefined(reading apply)"

### ISSUE THREAD

**Issue by sarulathadurai** [react-native-maps/react-native-maps#3955]
# Bug report

## Summary

I am trying to get the address onPress event by passing the coordinates to addressForCooordinate method but it throws error.

## Environment info

`react-native info` output:

```bash

      react-native: 0.64.2
      react: 17.0.1
      react-native-maps: ^0.29.0
```

## Code

```
          <MapView
          ref = {mapRef}
          provider={PROVIDER_GOOGLE} 
          style={[styles.map,{flex:1}]}
          showsUserLocation={true}
          followsUserLocation={true}
          loadingEnabled={true}
          showsMyLocationButton={true}
          mapType="standard"
          initialRegion={
              {
                latitude:position.latitude,
                longitude:position.longitude, 
                latitudeDelta: 0.0922,
                longitudeDelta: 0.0421
                }
            }
            onPress={getAddress}
        >
        <MapView.Marker
            coordinate={{latitude:position.latitude,longitude:position.longitude}}
            pinColor="red"
            draggable
            onDragEnd={(event)=>{setCoordinates(event.nativeEvent.coordinate)}}
        />

        </MapView>
```
```
 const getAddress = (e) => {
    mapRef.current.addressForCoordinate(e.nativeEvent.coordinate)
    .then((address) => {
      console.log('address', address); 
    }).catch((err) => {
      console.log('err', err); 
    })
  }
```
According to readme it should return promise with resolved address but
throws error as 
![Simulator Screen Shot - iPhone 12 - 2021-09-16 at 12 27 40](https://user-images.githubusercontent.com/52416786/133568184-5ebb4d6f-ffa5-4ecd-b33f-df109a1bdf4a.png)

**Comment by christopherdro** [comment]
@mechazod Any ideas?

**Comment by mechazod** [comment]
@sarulathadurai, please use `onRegionChangeComplete` https://github.com/react-native-maps/react-native-maps/issues/3911#issuecomment-903124543

**Comment by sarulathadurai** [comment]
@mechazod same error is thrown

**Comment by Simon-TechForm** [comment]
Could you please provide the entire example? What is your `mapRef`? What is `styles.map`? What is `position.latitude` and `position.longitude`? `What is your setCoordinates()`? What is your `position`?

**Comment by huangzichenzuishuaile** [comment]
@mechazod Have you solved this problem? I have the same problem now

**Comment by 00frank** [comment]
Nothing to do with this issue, but a fast run solution could be using a reverse geolocation api.
If you are developing this for a really small project or demo, you can use free Open Cage API and a the function could be exported like this, up to the devs add/merge/fix this package functionality.

```javascript
import env from "../env.json" // your private api keys / tokens

export default async function (coords) {
  let { latitude, longitude } = coords;
  let resp = await fetch(`https://api.opencagedata.com/geocode/v1/json?q=${latitude}+${longitude}&key=${env.geolocationServiceAPIKey}`)
  let data = await resp.json();

  return data.results[0] // this gets location info such as country, state, city, address
}
```

**Comment by andriy200017** [comment]
I have the same issue when using mapRef.current.addressForCoordinate inside the onPress callback for getting the address from coordinate. Have anyone got any solution for this to make it work on iOS? Using the another geo api is not a solution for me at all.

![Simulator Screen Shot - iPhone 13 - 2022-02-04 at 17 56 41](https://user-images.githubusercontent.com/40670081/152571363-c34374e8-0630-4c9a-8193-3e3738a9e843.png)

**Comment by danieldanielecki** [comment]
Not really, looks like a big issue :(

**Comment by danieldanielecki** [comment]
@bruno-de-queiroz can you share your code?
I've been debugging on `iOS` and same effect.

**Comment by Simon-TechForm** [comment]
This is more a missing feature and a lack of documentation, than a bug. `addressForCoordinate` isn't implemented for `provider={"google"}` on iOS.

@bruno-de-queiroz the following works on both Android and iOS, but not when using Google Maps on iOS.

```ts
import React, {useRef} from 'react';
import {StyleSheet} from 'react-native';
import MapView, {Region} from 'react-native-maps';

export default function App() {
  const mapRef = useRef<MapView>(null);

  const regionChangeCompleteHandler = async ({latitude, longitude}: Region) => {
    console.log(
      await mapRef.current?.addressForCoordinate({
        latitude,
        longitude,
      }),
    );
  };

  return (
    <MapView
      ref={mapRef}
      style={styles.mapView}
      onRegionChangeComplete={regionChangeCompleteHandler}
    />
  );
}

const styles = StyleSheet.create({
  mapView: {
    flex: 1,
  },
});
```

If anyone could open a pull request to clearly state in the docs that addressForCoordinate is Android and Apple Maps only, that would be extremely helpful.
Also, if anyone could put together a new issue as a feature request, requesting this functionality for Google Maps on iOS? That will leave us with a much better chance of somebody picking up the task.

**Comment by faycal-dev** [comment]
i solved this problem by saving e.nativeEvent.coordinate in a variable and then using it like this : 
onPress={(e) => {
            const pressCoords = e.nativeEvent.coordinate
            setLocation((prevState) => ({
              ...prevState,
              latitude: pressCoords.latitude,
              longitude: pressCoords.longitude,
            }));
          }}

**Comment by pejak4** [comment]
This isn't solved in version **react-native-maps": "^1.7.1**. Its working on Android but on iOS it still throws error **Cannot read properties of undefined (reading 'apply').**
